src/monkeyflocker/client.py

```diff
@@ -3,15 +3,36 @@
 """
 
 import asyncio
+import logging
 import math
+import os
+import pathlib
+import sys
 from dataclasses import InitVar, dataclass, field
-from typing import List
+from typing import Any, Dict, List, Optional
 
 import aiohttp
 import jinja2
+import structlog
+from structlog._config import BoundLoggerLazyProxy
 
 from .user import MonkeyflockerUser as MFUser
 
+DATE_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+
+@dataclass
+class ShadowRequest:
+    """This contains method, url, username, and body, if any.
+    Retrieving it from the actual request turns out to be tricky because
+    it's really a morass of little internal _*RequestContextManager objects.
+    """
+
+    method: str
+    url: str
+    username: str
+    data: Optional[str]
+
 
 @dataclass
```

Comment:
Also not sure why this is a data class, since you seem to have to go to some effort to make it not act like a data class (annotate fields to not be initialized, etc.).

Reply:
I guess on the whole I felt like I had more fields that WERE just set at __init__() than not?  the other ones are all constructing-the-user-list-from-base-name-base-uid-and-count and boilerplate for the logging.

And I like dataclasses and have been kind of fixated on them.  I guess I could construct the userlist in the cli and pass it to the class, but...ugh?